substrate/client/rpc/src/state/state_full.rs

```diff
@@ -165,6 +164,41 @@ where
 		}
 		Ok(())
 	}
+
+	// Prepares a storage diff from given start_keys and end_keys
+	fn prepare_storage_diff(
+		&self,
+		start : Block::Hash,
+		end : Block::Hash,
+		start_keys: Vec<StorageKey>,
+		end_keys: Vec<StorageKey>
+	)-> std::result::Result<Vec<(StorageKey, Option<StorageData>)>, JsonRpseeError> {
+		// prepare diff
+		let mut storage_diff : Vec<(StorageKey, Option<StorageData>)> = vec![];
+		for key in end_keys {
+			// if a new key is present, add it to the diff
+			if !start_keys.contains(&key) {
+				let new_storage = self.client.storage(end, &key).map_err(client_err)?;
+				storage_diff.push((key, new_storage))
+			} else {
+				let start_storage_val = self.client.storage(start, &key).map_err(client_err)?;
+				let end_storage_val = self.client.storage(end, &key).map_err(client_err)?;
+
+				if start_storage_val != end_storage_val {
+					storage_diff.push((key, end_storage_val))
+				}
+			}
+		}
+
+		// finally get any keys that have been removed between start_block and end_block
+		for key in start_keys {
+			if storage_diff.iter().position(|r| r.0 == key).is_none() {
+				storage_diff.push((key, None))
+			}
+		}
```

Comment:
It has a bug. You need to check `end_keys` instead of `storage_diff`.

```
for key in start_keys {
    if !end_keys.contains(&key) {
        storage_diff.push((key, None))
    }
}
```